docs/design/persistent-volume-provisioning.md

```diff
@@ -0,0 +1,140 @@
+### Goals
+
+* Allow administrators to describe minimum and maximum storage levels comprised of many kinds of PersistentVolumes
+* Allow the dynamic creation and reclamation of persistent volumes (to the fullest extent for each type of storage provider)
+
+
+#### API types
+
+```
+// PersistentVolumeController represents the configuration of a persistent volume.
+type PersistentVolumeController struct {
+	TypeMeta   `json:",inline"`
+	ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired specification of this PV controller.
+	Spec PersistentVolumeControllerSpec `json:"spec,omitempty"`
+
+	// Status is the current status of this PV controller.
+	Status PersistentVolumeControllerStatus `json:"status,omitempty"`
+}
+
+// PersistentVolumeControllerSpec is the specification of a PV controller.
+// The controller creates volumes exactly like the Template until the unbound count equals the replica count
+type PersistentVolumeControllerSpec struct {
+	// MinimumReplicas is the minimum number of unbound persistent volumes of this type desired in the system maintain
+	MinimumReplicas int `json:"minimumReplicas"`
+
+	// MaximumReplicas is the maximum total number of persistent volumes desired in the system
+	MaximumReplicas int `json:"maximumReplicas"`
+	
+	// Selector is a label query over persistent volumes which are managed by this controller
+	Selector map[string]string `json:"selector"`
+
+	// Template is the description of a PersistentVolume to create new replicas from
+	Template *PersistentVolumeTemplateSpec `json:"template,omitempty"`
+}
+
+// PersistentVolumeControllerStatus represents the current status of a PVController
+type PersistentVolumeControllerStatus struct {
+	// BoundReplicas is the number of replicas of this volume that are currently bound to PersistentVolumeClaims
+	BoundReplicas int `json:"boundReplicas"`
+	UnboundReplicas int `json:"unboundReplicas"`
+}
+
+// PersistentVolumeTemplateSpec describes the persistent volume created by this controller
+type PersistentVolumeTemplateSpec struct {
+	// Metadata of the PVs created from this template.
+	ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines a persistent volume.
+	Spec PersistentVolumeSpec `json:"spec,omitempty"`
+}
+
+```
+
+Dynamic pools of storage (e.g, GCE) will maintain `MinimumReplicas` of a `PersistentVolume` at a given time where each PVs status phase is ` Available`, up to `MaximumReplicas`.  Storage levels will increase by one until the minimum or maximum levels are reached.  Recycling a PV means the current number of available volumes might exceed the minimum replica count.  Plugins implementing the Deleter interface will delete PVs where the `PersistentVolumeReclaimPolicy` is set to Delete.  The default reclaim policy is Retain.
+
+Static pools of storage (e.g, NFS) will have ```MinimumReplicas``` == ```MaximumReplicas```.  Creation or deletion of the resource will not occur but the controller allows recycling of the volume through the plugin.
```

Comment:
Yes, the current recycler control loop will work just fine for the Delete reclaim policy.  I originally had it in there pre-1.0 but removed it until the rest of this design was ready to be implemented.

We need to add 1 controller for the PersistentVolumeControllerManager which will handle the PVCtrls and creation of PVs.

I'll update this doc to reflect the current design, too.  This proposal was made pre-1.0 and only recycling was able to fit at the very end.